src/bvals/cc/boundary_communication.cpp

```diff
@@ -0,0 +1,295 @@
+  Kokkos::Profiling::pushRegion("Task_LoadAndSendBoundBufs");
+
+  Mesh *pmesh = md->GetMeshPointer();
+  auto &cache = md->GetBvarsCache().GetSubCache(bound_type, true);
+
+  if (cache.buf_vec.size() == 0)
+    InitializeBufferCache<bound_type>(md, &(pmesh->boundary_comm_map), &cache, SendKey,
+                                      true);
+
+  auto [rebuild, nbound, other_communication_unfinished] =
+      CheckSendBufferCacheForRebuild<bound_type, true>(md);
+
+  if (nbound == 0) {
+    Kokkos::Profiling::popRegion(); // Task_LoadAndSendBoundBufs
+    return TaskStatus::complete;
+  }
+  if (other_communication_unfinished) {
+    Kokkos::Profiling::popRegion(); // Task_LoadAndSendBoundBufs
+    return TaskStatus::incomplete;
+  }
+
+  if (rebuild) RebuildBufferCache<bound_type, true>(md, nbound, BndInfo::GetSendBndInfo);
+
+  // Restrict
+  auto pmb = md->GetBlockData(0)->GetBlockPointer();
+  cell_centered_refinement::Restrict(cache.bnd_info, cache.bnd_info_h, pmb->cellbounds,
+                                     pmb->c_cellbounds);
+
+  // Load buffer data
+  const Real threshold = Globals::sparse_config.allocation_threshold;
+  auto &bnd_info = cache.bnd_info;
+  PARTHENON_DEBUG_REQUIRE(bnd_info.size() == nbound, "Need same size for boundary info");
+  auto &sending_nonzero_flags = cache.sending_non_zero_flags;
+  auto &sending_nonzero_flags_h = cache.sending_non_zero_flags_h;
+  Kokkos::parallel_for(
+      "SendBoundBufs",
+      Kokkos::TeamPolicy<>(parthenon::DevExecSpace(), nbound, Kokkos::AUTO),
+      KOKKOS_LAMBDA(parthenon::team_mbr_t team_member) {
+        const int b = team_member.league_rank();
+
+        sending_nonzero_flags(b) = false;
+        if (!bnd_info(b).allocated) return;
+
+        const int &si = bnd_info(b).si;
+        const int &ei = bnd_info(b).ei;
+        const int &sj = bnd_info(b).sj;
+        const int &ej = bnd_info(b).ej;
+        const int &sk = bnd_info(b).sk;
+        const int &ek = bnd_info(b).ek;
+        const int Ni = ei + 1 - si;
+        const int Nj = ej + 1 - sj;
+        const int Nk = ek + 1 - sk;
+
+        const int &Nt = bnd_info(b).Nt;
+        const int &Nu = bnd_info(b).Nu;
+        const int &Nv = bnd_info(b).Nv;
+
+        const int NjNi = Nj * Ni;
+        const int NkNjNi = Nk * NjNi;
+        const int NvNkNjNi = Nv * NkNjNi;
+        const int NuNvNkNjNi = Nu * NvNkNjNi;
+        const int NtNuNvNkNjNi = Nt * NuNvNkNjNi;
+
+        Kokkos::parallel_for(Kokkos::TeamThreadRange<>(team_member, NtNuNvNkNjNi),
+                             [&](const int idx) {
+                               const int t = idx / NuNvNkNjNi;
+                               const int u = (idx % NuNvNkNjNi) / NvNkNjNi;
+                               const int v = (idx % NvNkNjNi) / NkNjNi;
+                               const int k = (idx % NkNjNi) / NjNi + sk;
+                               const int j = (idx % NjNi) / Ni + sj;
+                               const int i = idx % Ni + si;
```

Comment:
The difference here is that we need both the flat index and the shaped indices to copy `var` into `buf`. I had in mind implementing something like [`numpy.unravel_index`](https://numpy.org/doc/stable/reference/generated/numpy.unravel_index.html). However, if we naively implemented that it we'd be recomputing the intermediate integers on each vector lane instead of once per thread.